### PR TITLE
[EXT-103] Zendesk

### PR DESCRIPTION
**Pull request by michal-grzelak**
Add Zendesk extensions with actions:
- Create task
- Complete task

Tested with real data.

### CODE REVIEW

File: extensions/zendesk/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+# Zendesk
+
+Zendesk is a customer service and support platform that provides a suite of tools and software for businesses to manage their customer interactions. It offers a range of customer service solutions, including help desk ticketing systems, customer support and engagement, and self-service options. Zendesk is designed to help companies streamline their customer support processes, improve communication with their customers, and resolve issues more efficiently.
+
+## Extension settings
+
+In order to set up this extension, you will need:
+
+2. An **Sales API token** - visit Zendesk Sales CRM dashboard and go to **Settings > Integration > OAuth** to enable and create API token
```
Comment (nckhell) [inline extensions/zendesk/README.md:14]:
```suggestion
In order to set up this extension, you will need a **sales API token**. You can visit Zendesk Sales CRM dashboard and go to **Settings > Integration > OAuth** to enable and create an API token.
```

File: extensions/zendesk/settings.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import { z, type ZodTypeAny } from 'zod'
+
+export const settings = {
+  salesApiToken: {
+    label: 'Sales API token',
+    key: 'salesApiToken',
+    obfuscated: true,
+    required: true,
+    description:
+      'Visit Zendesk Sales CRM dashboard and go to Settings > Integration > OAuth to enable and create API token',
```
Comment (nckhell) [inline extensions/zendesk/settings.ts:11]:
```suggestion
      'Visit Zendesk Sales CRM dashboard and go to Settings > Integration > OAuth to enable and create an API token',
```

File: extensions/zendesk/v1/actions/completeTask/config/fields.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  NumericIdSchema,
+} from '@awell-health/extensions-core'
+
+export const fields = {
+  taskId: {
+    label: 'Task ID',
+    id: 'taskId',
+    type: FieldType.NUMERIC,
+    required: true,
+    description: '',
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/completeTask/config/fields.ts:14]:
```suggestion
    description: 'The ID of the task to complete in Zendesk',
```

File: extensions/zendesk/v1/actions/createTask/config/fields.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    required: true,
+    description: 'Content of the task.',
+  },
+  dueDate: {
+    label: 'Due date',
+    id: 'dueDate',
+    // Time will be auto-set to 00:00:00 as FieldType.DATE provides only date
+    type: FieldType.DATE,
+    required: false,
+    description: "Date of task's due date in UTC (ISO8601 format).",
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/createTask/config/fields.ts:27]:
```suggestion
    description: "",
```

File: extensions/zendesk/v1/actions/createTask/config/fields.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  },
+  ownerId: {
+    label: 'Owner ID',
+    id: 'ownerId',
+    type: FieldType.NUMERIC,
+    required: false,
+    description:
+      'Unique identifier of the user the task is assigned to. Defaults to the unique identifier of the user who created the task.',
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/createTask/config/fields.ts:35]:
```suggestion
      'Unique identifier of the user the task is assigned to, defaults to the user who created the task',
```

File: extensions/zendesk/v1/actions/createTask/config/fields.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    description: 'Indicator of whether the task is completed or not.',
+  },
+  remindAt: {
+    label: 'Remind at',
+    id: 'remindAt',
+    // Time will be auto-set to 00:00:00 as FieldType.DATE provides only date
+    type: FieldType.DATE,
+    required: false,
+    description:
+      'Date that we should send you a reminder about the task in UTC (ISO8601 format).',
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/createTask/config/fields.ts:66]:
```suggestion
      'The date at which Zendesk should send you a reminder about the task',
```

File: extensions/zendesk/v1/actions/createTask/config/fields.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+const resourceTypeEnum = z.enum<
+  ResourceType,
+  [ResourceType, ...ResourceType[]]
+>([ResourceType.CONTACT, ResourceType.DEAL, ResourceType.LEAD])
+
+export const FieldsValidationSchema = z.object({
+  content: z.string().nonempty(),
+  dueDate: DateTimeOptionalSchema,
+  ownerId: makeStringOptional(NumericIdSchema),
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/createTask/config/fields.ts:78]: Just double-checking whether this is correct: `makeStringOptional(NumericIdSchema)`

ownerId is a number, as indicated by `NumericIdSchema` but then you wrap it in `makeStringOptional` which explicitly says "make this string optional" (while it's a number).

File: extensions/zendesk/v1/actions/createTask/config/fields.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  resourceType: makeStringOptional(resourceTypeEnum),
+  resourceId: makeStringOptional(NumericIdSchema),
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/createTask/config/fields.ts:80]: Same comment here as with `ownerId`.

File: extensions/zendesk/v1/actions/createTask/createTask.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import {
+  ZendeskClient,
+  isSalesApiError,
+  salesApiErrorToActivityEvent,
+} from '../../client'
+
+export const createTask: Action<typeof fields, typeof settings> = {
+  key: 'createTask',
+  title: 'Create task',
+  description: 'Creates a new task',
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/createTask/createTask.ts:16]:
```suggestion
  description: 'Creates a new task in Zendesk',
```

File: extensions/zendesk/v1/actions/completeTask/completeTask.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import {
+  ZendeskClient,
+  isSalesApiError,
+  salesApiErrorToActivityEvent,
+} from '../../client'
+
+export const completeTask: Action<typeof fields, typeof settings> = {
+  key: 'completeTask',
+  title: 'Complete a task',
+  description: 'Sets task completed status to true',
```
Comment (nckhell) [inline extensions/zendesk/v1/actions/completeTask/completeTask.ts:16]:
```suggestion
  description: 'Completes a task in Zendesk',
```

**Review by nckhell — CHANGES_REQUESTED** [review]

File: extensions/zendesk/v1/actions/createTask/config/fields.ts (hunk 7 above)
Comment (michal-grzelak) [reply extensions/zendesk/v1/actions/createTask/config/fields.ts:78]: Yes, I think `makeStringOptional` should be named a little different, but it seems that's the best I could come up at the time I was making it. Under the hood this utility function is `z.union([schema.optional(), z.literal('').transform(() => undefined)])`, which means that you can pass any schema there and it just marks a schema as optional and makes sure that if empty string `''` is passed it is still `undefined` (e.g. situation when you set sth in Awell, but then clear the input)

**Review by nckhell — APPROVED** [review]

**Pull request merged**